### DISCUSSION TITLE
BLOB streaming API

### DISCUSSION THREAD

**Discussion by fschulze** [sqlalchemy/sqlalchemy#10549] [Ideas]
I would be interested in implementing BLOB streaming support for pg8000, sqlite3 and maybe psycopg3. I guess to explore the space it would be best to do it as an SQLAlchemy addon first. Do you have any suggestions on how to get started with that? Which parts should I look/hook into? Are there similar things, like the dialect stuff, that are already in SQLAlchemy or an addon where I can look to see how I could add that? How could an abstract but optional API look like? There needs to be some kind of feature detection for sqlite3 on older Python versions for example. So far I only used dialect stuff for applications with only one specific database backend.

- For sqlite3 the support was added with Python 3.11: https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.blobopen
- For pg8000 there is this: https://pypi.org/project/pg8000/#copy-from-and-to-a-stream
- For psycogp3 there is this: https://www.psycopg.org/psycopg3/docs/basic/copy.html
- It looks like Oracle and SQL Server also support blob streaming, but I don't have access to those

**Comment by CaselIT** [comment]
Hi,

While maybe something like this may make sense, I'm not sure the API are similar enough for a solution that's not custom for each driver.

For example looking at the sqlite api, I'm sure how it should be used.

> table ([str](https://docs.python.org/3/library/stdtypes.html#str)) – The name of the table where the blob is located.
> column ([str](https://docs.python.org/3/library/stdtypes.html#str)) – The name of the column where the blob is located.
> row ([str](https://docs.python.org/3/library/stdtypes.html#str)) – The name of the row where the blob is located.

What is the name of the row?

**Comment by zzzeek** [comment]
Oracle's streaming API is the best and it's the one you should be looking at for how this would be adapted to other dialects.   You run a SELECT, you get rows.  then, inside the columns that have BLOBs, the data value you get back is a Python file object which you stream.  Perfect!  What are the limitations?  Well you have to stream that entire blob as you get the row, if you keep reading rows and try to put all the filehandles someplace to get them later, they might not work if the result set has paged to the next batch.  also there might be restrictions with rows that have more than one BLOB column per row. 

the API for this is at https://python-oracledb.readthedocs.io/en/latest/user_guide/lob_data.html#streaming-lobs-read.   SQLAlchemy turns this API off by default (see https://docs.sqlalchemy.org/en/20/dialects/oracle.html#lob-datatypes).   From the SQLAlchemy side, turning on /off BLOB streaming could be something based on using [server side results](https://docs.sqlalchemy.org/en/20/core/connections.html#engine-stream-results) in conjunction with an execution option.


now, all of that above is kind of an "ideal".   In reality.    These APIs are all enomously different.  the PG COPY thing I dont even understand how one targets specific columns/rows in a table and I'm not even sure that's the appropriate API for PG.    The reality is, people stopped caring about streaming BLOBs for two reasons:  1. computers started having more than 16 megs of RAM in the mid-90s 2. it's generally considered poor form to store enormous BLOBs in a relational database; large files are usually stored as just that, files, on a file system or specialized document database.

To that end, an agnostic "stream blobs from the database" API I'd just implement in the most simple way possible to suit the current application's needs and the target backends they support.  As far as SQLAlchemy, i'd skip it entirely and [use the cursor from each backend to do what's necessary to get the work done](https://docs.sqlalchemy.org/en/20/core/connections.html#working-with-the-dbapi-cursor-directly).    There would not be a pypi package; it would just be implemented as needed in the application.

**Reply by fschulze** [reply]
Glad I asked first. I will do it as app specific code then and maybe blog about it afterwards. Thanks for the pointers!
